README.md

```diff
@@ -1,2 +1,202 @@
+	NewReader() (io.ReadCloser, error)
+}
+
+// streamer structure embeds os.File for implementing
+// simultaneous output streaming for multiple clients.
+type streamer struct {
+	*os.File // embeds os.File
+}
+
+```
+  Note that `streamer` structure embeds `os.File` and has many benefits:
+  1. Read() and Write() methods for free
+  2. Synchronization by kernel for free
+  3. Memory-error safe as data is stored in disk
+  4. Will be asigned to exec.Cmd.Stdout and exec.Cmd.Stderr fields and both outputs will go into a single file in chronological order.
+
+Say, if a client is requesting for a job that builds a project, logs can be very long and may take up a lof of space. The above approach is very handy as it uses disk space rather than in-memory buffers. And implementation will be simple as many things come for free.
+A disadvantage to this approach is that it requires some effort to make the disk space available at all times. A cron job can be used to remove files with certain age from directory where all job outputs are collected.
+
+Server (as library user) will track all jobs created in-memory via a map data structure. To make it more practicle, other options can be used (database, file system, infra provided by kernel, etc.)
+
+### Using library
+Following are the functions this library will provide
+```go
+// NewJob creates a job with specified command but does not start it
+func NewJob(command string) *Job
+// Start starts the job execution. Calls exec.Cmd.Start() implicitly
+// If the call fails, its error is returned by Start()
+func (j* Job) Start() error {}
+// Status returns status of this job. See struct JobStatus for more details
+// Users can determine whether specified job has completed or not via this function
+func (j *Job) Status() JobStatus {}
+// Stop terminates a running job. Does nothing if job has already stopped.
+func (j *Job) Stop() error {}
+// Output returns a string channel which will be closed by library internally
+// once the job completes. Multiple calls to this function will create separate channels
+// so it is safe to stream output to multiple clients.
+func (j *Job) Output() chan string {}
+// Pid returns PID of process running the job.
+// If no such process exist, an error will be returned.
+func (j *Job) Pid() (int, error) {}
+```
+
+## gRPC
+To implement connectivity between clients and the server, gRPC package will be used.
+Following proto file will be used for RPCs.
+```go
+service Worker {
+  // Starts a job
+  rpc Start(StartRequest) returns (StartResponse) {}
+  // Stops the specified job
+  rpc Stop(StopRequest) returns (StopResponse) {}
+  // Gets status of a job
+  rpc Status(StatusRequest) returns (StatusResponse) {}
+  // Get output of a job. If job is still running, it will stream
+  // the output
+  rpc Output(OutputRequest) returns (stream OutputResponse) {}
+}
+
+message StartRequest {
+  string command = 1;
+}
+
+message StartResponse {
+  string command = 1;
```

Comment:
If you need the job ID in order to stop a job or get its status, shouldn't the start response include the ID of the job that was started?

Reply:
It actually does return job id. I don't know why this part of document has stale data.
Below is the current proto I am using. Will update this document to reflect exact same.

https://github.com/ohasanliUW/simple-worker/blob/client_server/protobuf/worker.proto